processing/src/main/java/org/apache/druid/query/operator/SegmentToRowsAndColumnsOperator.java

```diff
@@ -36,33 +39,22 @@ public SegmentToRowsAndColumnsOperator(
   }
 
   @Override
-  public void open()
+  public void go(Receiver receiver)
   {
-
-  }
-
-  @Override
-  public RowsAndColumns next()
-  {
-    hasNext = false;
-
-    RowsAndColumns rac = segment.as(RowsAndColumns.class);
-    if (rac != null) {
-      return rac;
+    try (final CloseableShapeshifter shifty = segment.as(CloseableShapeshifter.class)) {
```

Comment:
it strikes me as a bummer that the segment thing you get here isn't already a `CloseableShapeshifter` to manage the lifecycle of the `RowsAndColumns` to be materialized from it.

I like the `CloseableShapeshifter` pattern, I think maybe we could apply it elsewhere to improve our handling of closeables, because its a bit of a dizzying mess in `Segment` which is why I think this bothers me a bit.

No need to change anything now, just something I want to think about a bit

Reply:
I agree that it's a dizzying mess in `Segment`.  Hopefully, this contributes to eventually making it a less-dizzying mess in some future world.